examples/rlhf/train_rlhf.py

```diff
@@ -0,0 +1,385 @@
+                        row,
+                        EOS_TOKEN_ID,
+                    )
+                    # + 1 so that we get prompt and full generated sequence as first
+                    # and last row respectively
+                    for i in range(max_new_tokens + 1)
+                ]
+            )
+            for rindex, row in zip(batch.prompt_rindex, generated)
+        ],
+    )
+    rollout_attention_mask = (rollout_generated != EOS_TOKEN_ID).to(torch.int64)
+
+    # done is True when we either first sample an EOS token or reach the maximum number
+    # of generated tokens
+    done_idx = torch.minimum(
+        (generated != EOS_TOKEN_ID).sum(dim=-1) - batch.prompt_rindex,
+        torch.tensor(max_new_tokens),
+    )
+    done = (
+        torch.arange(max_new_tokens, device=generated.device) == done_idx[:, None]
+    ).unsqueeze(-1)
+
+    # the sequence of actions for each trajectory is just the generated token ids
+    action_idx = torch.stack(
+        [
+            torch.arange(i, i + max_new_tokens, device=generated.device)
+            for i in batch.prompt_rindex
+        ]
+    )
+    action = generated[
+        torch.arange(generated.shape[0], device=generated.device)[:, None],
+        action_idx,
+    ]
+
+    # calculate the reward for the finished sequence
+    _, end_scores = reward_model(
+        input_ids=rollout_generated[:, -1], attention_mask=rollout_attention_mask[:, -1]
+    )
+    _, end_scores_labels = reward_model(
+        input_ids=batch.input_ids,
+        attention_mask=batch.attention_mask,
+    )
+    # TODO: add KL penalty in reward
+    # the reward is zero except for the timestep where we reached a stopping condition
+    reward = (
+        done * (end_scores - end_scores_labels)[:, None, None]
+        - kl_coef * log_ratio[..., None]
+    )
+    td = {
+        "action": action,
+        "input_ids": rollout_generated[:, :-1].clone(),
+        "attention_mask": rollout_attention_mask[:, :-1].clone(),
+        "sample_log_prob": log_probs,
+        "next": {
+            "input_ids": rollout_generated[:, 1:].clone(),
+            "attention_mask": rollout_attention_mask[:, 1:].clone(),
+            "done": done,
+            "reward": reward,
+        },
+    }
+    return TensorDict(td, batch_size=done.shape[:2], device=generated.device)
+
+
+def flatten_td(td):
+    # our tensordict has shape [B, T] where B = batch_size and T = trajectory length
+    # some trajectories may have stopped (reached EOS) before generating T tokens
+    # this function truncates and concatenates the trajectories, resulting in a
+    # tensordict that has shape [N] where N <= B * T.
+    done = td["next", "done"]
+    mask = torch.zeros_like(done)
+    mask[..., 1:, :] = done[..., :-1, :]  # shift by one
+    mask = ~mask.cumsum(-2).bool().squeeze()
+    return td[mask]
+
+
+def create_loss_estimator(
+    config, reward_model, batch, tokenizer, logger=None, ref_model=None
+):
```

Comment:
A docstring would be nice!

Reply:
This seems to be returning a local function, I'm not sure I understand what it is doing

Reply:
Yes, it's basically a factory function that creates an `estimate_loss` function that we call in the training loop to evaluate the model. Perhaps would be more idiomatic to use a class with a `__call__` method rather than a functional closure.